### PR TITLE
 Convert TravisCi tests to a GithubActions workflow

### PR DESCRIPTION
**Pull request by VadimSchmitz**
This pull request converts Stepup-tiqr from Homestead and TravisCI to Docker and GitHub Actions.

work done to make this conversion:
- Add docker functionality with docker-compose.
- Refactor composer.json so each qa-test has their own bash scripts with configs in a different folder.
- Replace TravisCi with a GitHub actions workflow.
- Replace non functional phpcpd with jscpd.
- Update testing.md where needed.

### CODE REVIEW

File: ci/files/.gitignore [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# Ignore everything in this directory
+*
+# Except this file
+!.gitignore
```
Comment (MKodde) [inline ci/files/.gitignore:4]: :muscle:

File: .github/workflows/test-integration.yml [hunk 2]
```diff
@@ -0,0 +1,52 @@
+name: test-integration
+on:
+  pull_request:
+  push:
+#    branches: [ master, develop ]
```
Comment (MKodde) [inline .github/workflows/test-integration.yml:5]: Dev leftover?

File: .github/workflows/test-integration.yml [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+          cp .env.ci .env
+          cp config/legacy/parameters.yaml.dist config/legacy/parameters.yaml
+          '
+      - name: Build frontend assets
+        run: cd ci/docker && docker-compose exec -T tiqr.php.test.vm.openconext.org bash -lc 'yarn encore production'
+      - name: Run test scripts
+        run: cd ci/docker && docker-compose exec -T tiqr.php.test.vm.openconext.org bash -lc ' composer test '
+      - name: Output log files on failure
+        if: failure()
+        run: cd ci/docker && docker-compose exec -T tiqr.php.test.vm.openconext.org -c 'cat var/log/webtest.log'
```
Comment (MKodde) [inline .github/workflows/test-integration.yml:52]: I don't think the webtest file will exist. See feedback on the previous PR's where we had similar logging challenges. And make this line uniform with the other.

File: .github/workflows/test-integration.yml [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+      - name: Init environment
+        run: |
+          cd ci/docker
+          docker-compose up -d
+      - name: Install dependencies
+        run: |
+          cd ci/docker && docker-compose exec -T tiqr.php.test.vm.openconext.org bash -lc '
+          composer install
+          yarn
+          cp .env.ci .env
```
Comment (MKodde) [inline .github/workflows/test-integration.yml:43]: Have you seen the contents of the `.env.ci`. The Symfony env is set to 'dev'. Can you check why this is the case? And (if possible) migrate this to run in the `test` environment.

File: .github/workflows/test-integration.yml [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+      - uses: actions/cache@v2
+        id: yarn-cache
+        with:
+          path: ${{ steps.yarn-cache-dir-path.outputs.dir }}
+          key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}
+          restore-keys: |
+            ${{ runner.os }}-yarn-
+      - name: Generate SSL key and certificate
+        run: |
+          sudo openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout ./ci/files/tiqr.example.com.key -out ./ci/files/tiqr.example.com.crt -subj "/C=NL/ST=Netherlands/L=Amsterdam/O=TEST/CN=tiqr.example.com"
```
Comment (MKodde) [inline .github/workflows/test-integration.yml:33]: Turn this into a oneliner?

File: docs/testing.md [hunk 6]
```diff
@@ -59,9 +59,11 @@ Acceptance tests will be written for all command handlers and will be run as par
 
 You can run these tools manually with composer:
 
+> **Notice**: The behat tests sometimes gives a false positive, if the tests fail retry the tests before troubleshooting
```
Comment (MKodde) [inline docs/testing.md:62]: Muchios gracias for adding this warning!

Proposal to make this sentence more specific:
old: `, if the tests fail retry the tests before troubleshooting`
new: `. If the <<inser test name here>> test fails, please re-run the behat tests before troubleshooting`

File: ci/qa/static-analysis [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+./vendor/bin/phpmd src text ci/qa/qa-config/phpmd.xml --suffixes php
+echo "phpmd static analysis passed"
+
+#jscpd static analysis
+# Run the JS CPD tool:
+# - ignoring the test files and public resources
+# - allowing for 40 lines of duplicate code
+# - and setting a total copy paste limit to 1% project wide
+# More info about jscpd usage: https://github.com/kucherenko/jscpd/tree/master/packages/jscpd#usage
+./node_modules/.bin/jscpd --ignore "**/Tests/*.php,**/Resources/public/**/*" src  -l 40 -t 1
```
Comment (MKodde) [inline ci/qa/static-analysis:19]: Please review this jscpd call: 
- We miss out on testing our TS code
- I have not found any Unit tests so far, maybe remove the ignore on the tests?
- Tiqr does not adhere to the SF 3 bundle based setup, the ignore section does assume it does. :man_shrugging: 
- Maybe run this as the final static analysis task. Failing on CPD is less informative than having a PHP CS issue in my opinion.

**Review by MKodde — CHANGES_REQUESTED** [review]
Nice work again @VadimSchmitz 

Some feedback can be found below in the code review. 

In addition:

1. You might have been to quick in removing the Homestead setup?
2. If you do decide to remove Homestead, than please document this and remove any left-over references.

**Bonus commit message feedback**
Thanks for creating more atomic commits. Keep it up! And I would encourage you to look at your work with creating that kind of commits in mind. Now you rewrote the log at the end, right before creating the PR. And some commits show this. For example, the static analysis task runner was not added on 4e1ce6a (the composer test call was tho), but was introduced in: e9583871e221a48171816a5bb0146e10560a0b7f. Small details that make reviewing somewhat harder. But all in all you did a stellar job!

File: ci/qa/static-analysis (hunk 7 above)
Comment (MKodde) [reply ci/qa/static-analysis:19]: On a broader note. The output of the different static analysis calls looks kind of a mess. Would you be so kind to add an echo statement for each tool that is run?

`printf "\nRunning PHP Code Style\n"` Something along those lines.

File: .github/workflows/test-integration.yml (hunk 2 above)
Comment (VadimSchmitz) [reply .github/workflows/test-integration.yml:5]: Missed that one, It has been uncommented

File: .github/workflows/test-integration.yml (hunk 3 above)
Comment (VadimSchmitz) [reply .github/workflows/test-integration.yml:52]: Resolved it the same way as [Stepup-Middleware](https://github.com/OpenConext/Stepup-Middleware)

File: .github/workflows/test-integration.yml (hunk 4 above)
Comment (VadimSchmitz) [reply .github/workflows/test-integration.yml:43]: The tests should indeed use the test env instead over the dev env. It has now been changed to test.

File: .github/workflows/test-integration.yml (hunk 5 above)
Comment (VadimSchmitz) [reply .github/workflows/test-integration.yml:33]: Done

File: ci/qa/static-analysis (hunk 7 above)
Comment (VadimSchmitz) [reply ci/qa/static-analysis:19]: phpcpd will currently not work due to version constraints.
regarding your points.
- jscpd arguments have been updated to also check typescript and scss files.
- Tests folder has been removed for now.
- the ignore section was not needed.
- jscpd has been moved to the bottom of the static-tests list.

File: .env.ci [hunk 8]
```diff
@@ -14,7 +14,7 @@
 # https://symfony.com/doc/current/best_practices/configuration.html#infrastructure-related-configuration
 
 ###> symfony/framework-bundle ###
-APP_ENV=dev
+APP_ENV=test
```
Comment (MKodde) [inline .env.ci:17]: :+1:

File: .github/workflows/test-integration.yml (hunk 4 above)
Comment (MKodde) [reply .github/workflows/test-integration.yml:42]: Good of you to add the cp .env command. In order to let it work without side effects, I suspect you'll need to move that command to a point before running composer install (as that will trigger Symfony scripts, which in turn need to run in the correct environment).

The same goes for the copying of the legacy parameters.

File: .github/workflows/test-integration.yml [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+          cp .env.ci .env
+          cp config/legacy/parameters.yaml.dist config/legacy/parameters.yaml
+          '
+      - name: Build frontend assets
+        run: cd ci/docker && docker-compose exec -T tiqr.php.test.vm.openconext.org bash -lc 'yarn encore production'
+      - name: Run test scripts
+        run: cd ci/docker && docker-compose exec -T tiqr.php.test.vm.openconext.org bash -lc ' composer test '
+      - name: Output log files on failure
+        if: failure()
+        run: tail -2000 /var/log/syslog
```
Comment (MKodde) [inline .github/workflows/test-integration.yml:51]: :100:

File: .gitignore [hunk 10]
```diff
@@ -76,3 +76,6 @@ yarn-error.log
 ###< symfony/phpunit-bridge ###
 
 local-php-security-checker
+*.crt
+*.key
```
Comment (MKodde) [inline .gitignore:80]: I can imagine (test) scenarios where we want to keep a cert or even a private key in vcs. Having this rigid rule in place will make it harder to add them. For now, feel free to leave this as is. But a good rule of thumb is to be specific in .gitignore files. Not being to loose in excluding things.

File: ci/qa/behat [hunk 11]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
+
+cd $(dirname $0)/../../
+
+vendor/bin/behat --stop-on-failure --config ./ci/qa/qa-config/behat.yml --tags=~skip
+echo "Behat tests passed"
```
Comment (MKodde) [inline ci/qa/behat:6]: I personally prefer a pre run statement (with newlines to visually distinguish the output of the previous run with that of the current output. 

For that you need to use `printf` as `echo` does not allow adding linefeeds `\n`

This feedback also goes for the other places where you added a 'post-run' report.

File: ci/qa/lint [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+
+cd $(dirname $0)/../../
+
+#Lints PHP,yml,ts,tsc and composer.json
+vendor/bin/parallel-lint dev src
+bin/console lint:yaml config
+yarn tslint --project tsconfig.json
+yarn tsc --noEmit
+composer validate
+echo "Application Linting passed"
```
Comment (MKodde) [inline ci/qa/lint:11]: :point_up:

File: ci/qa/phpunit [hunk 13]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env bash
+
+cd $(dirname $0)/../../
+
+XDEBUG_MODE=coverage ./vendor/bin/phpunit --configuration=ci/qa/qa-config/phpunit.xml --coverage-text
```
Comment (MKodde) [inline ci/qa/phpunit:5]: And not a single PHP Unittest was run that day :rofl:

File: docs/testing.md [hunk 14]
```diff
@@ -59,9 +59,11 @@ Acceptance tests will be written for all command handlers and will be run as par
 
 You can run these tools manually with composer:
 
+> **Notice**: The behat test: [registration.feature](https://github.com/OpenConext/Stepup-tiqr/blob/develop/src/Features/registration.feature)
+sometimes gives a false positive, if the tests fail retry the tests before troubleshooting.
```
Comment (MKodde) [inline docs/testing.md:63]: :bow:

**Review by MKodde — COMMENTED** [review]
Awesome work so far! Here's an intermediate review. I konw you are still working on the Homestead stuff.

**Comment by VadimSchmitz** [comment]
The build is currently failing due to symfony/webpack-encore having a high risk vulnerability.
This will need to be solved in another PR.

File: .github/workflows/test-integration.yml [hunk 15]
```diff
@@ -0,0 +1,51 @@
+name: test-integration
+on:
+  pull_request:
+  push:
+#    branches: [ master, develop ]
```
Comment (MKodde) [inline .github/workflows/test-integration.yml:5]: Dev leftover?

File: ci/docker/init.sh [hunk 16]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env bash
+uid=$(id -u)
+gid=$(id -g)
+
+printf "UID=${uid}\nGID=${gid}\nCOMPOSE_PROJECT_NAME=eb" > .env
```
Comment (MKodde) [inline ci/docker/init.sh:5]: COMPOSE_PROJECT_NAME is set to: `eb`. That does not make sense right? Is the init.sh script still used?

File: src/Features/authentication.feature [hunk 17]
```diff
@@ -34,45 +34,45 @@ Feature: When an user needs to authenticate
     Then I should see "urn:oasis:names:tc:SAML:2.0:status:Success"
 
     And the logs are:
-      | level  | message                                                                                                                         | sari    |
+      | level  | message                                                                                                                                 | sari    |
```
Comment (MKodde) [inline src/Features/authentication.feature:37]: The majority of the changed lines are only reformatting the file. This obfuscates the actual change. Can you revert the styling? Or make it a 2 commit business?

File: src/Features/registration.feature [hunk 18]
```diff
@@ -23,50 +23,50 @@ Feature: When an user needs to register for a new token
     Then I should see "urn:oasis:names:tc:SAML:2.0:status:Success"
 
     And the logs are:
-      | level  | message                                                                                                                       | sari    |
+      | level  | message                                                                                                                              | sari    |
```
Comment (MKodde) [inline src/Features/registration.feature:26]: Idem here. Please revert the re-formatting :bow:

**Review by MKodde — CHANGES_REQUESTED** [review]
Almost there. See some feedback in the code review section. Not all of the previous review points where addressed. I left the ones open that might need attention.

File: .github/workflows/test-integration.yml (hunk 4 above)
Comment (VadimSchmitz) [reply .github/workflows/test-integration.yml:42]: Copying does now happen before the composer install call

File: .gitignore (hunk 10 above)
Comment (VadimSchmitz) [reply .gitignore:80]: It has been changed to specifically ignore the tiqr.stepup.example.com key and crt

File: ci/qa/behat (hunk 11 above)
Comment (VadimSchmitz) [reply ci/qa/behat:6]: echo has been replaced by printf where it was used, I will remember this for the next time i need to ouput text.

File: .github/workflows/test-integration.yml (hunk 15 above)
Comment (VadimSchmitz) [reply .github/workflows/test-integration.yml:5]: has been uncommented

File: src/Features/authentication.feature (hunk 17 above)
Comment (VadimSchmitz) [reply src/Features/authentication.feature:37]: This has now been split up into two commits

File: src/Features/registration.feature (hunk 18 above)
Comment (VadimSchmitz) [reply src/Features/registration.feature:26]: see https://github.com/OpenConext/Stepup-tiqr/pull/143#discussion_r852874098

File: ci/qa/behat (hunk 11 above)
Comment (MKodde) [reply ci/qa/behat]: And please printf before running the test. Like we report on the other tasks (we advertise a task is about to run).

Comment (VadimSchmitz) [reply ci/qa/behat]: I have changed it into a start message instead of a completed message.

File: composer.json [hunk 19]
```diff
@@ -102,6 +102,9 @@
     "sort-packages": true,
     "allow-plugins": {
       "symfony/flex": true
+    },
+    "platform": {
+      "php": "7.2.5"
```
Comment (MKodde) [inline composer.json:107]: Nice addition!

Also, feel free to add `"php": "^7.2",` as a dependency. As we do in EngineBlock: 

https://github.com/OpenConext/OpenConext-profile/blob/a3aa579fa1565bb76c3693e246c18e8e572e8b7a/composer.json#L14

But feel free to leave this as-is. For now this is good enough.

File: .github/workflows/test-integration.yml (hunk 15 above)
Comment (MKodde) [inline .github/workflows/test-integration.yml:5]: Three times' a charm

Comment (VadimSchmitz) [reply .github/workflows/test-integration.yml:5]: classic, it has now been changed

File: composer.json (hunk 19 above)
Comment (VadimSchmitz) [reply composer.json:107]: I will leave it for now.

**Review by MKodde — APPROVED** [review]
Ship-it :ship:

File: CHANGELOG.md [hunk 20]
```diff
@@ -1,3 +1,7 @@
+## 3.4.0
```
Comment (MKodde) [inline CHANGELOG.md:1]: Given this is not realy a user-facing feature I'd opt for a patch release, 3.3.1

**Pull request merged**